New webhook for adding dco-check webhook to newly created/publicized repos

robbkidd opened the pull request:
Depends on the changes in #6 for the relocation of the gem files into `src/` because cleanliness is next to and all that.

robbkidd commented:
For you, @tas50. I plan to spin a test instance of this up and point a personal play-toy organization at it.

robbkidd commented on src/lib/dcob.rb:
```diff
@@ -86,6 +104,11 @@ def verify_signature(payload_body)
+  if config["cfg"]["dco_webhook_url"]
+    DCO_WEBHOOK_URL = config["cfg"]["dco_webhook_url"]
+  else
+    puts "You must specify cfg.dco_webhook_url in config.toml"
```
I wonder if this really needs to be configurable. The webhook URL (e.g. "me" from the bot's perspective) could be determined from the incoming `request.url` or `request.scheme + '://' + request.host + '/payload'`

robbkidd commented on src/lib/dcob.rb:
```diff
@@ -1,49 +1,66 @@
+    post "/payload", event_type: "repository" do
+      begin
+        repo = JSON.parse(@payload_body)
+        if %w{created publicized}.include? repo["action"]
```
Also filter on only public repos.

robbkidd committed:
add a new payload variant for repository events

* Create a hookit method and place it in a GitHub client utility class
* Class method added to Octoclient to make calling hookit read better
* Figure out webhook to add based on incoming request URL
* Add TESTS!
* Use chefstyle and make our world better

Because these webhooks are signed, we can trust that the incoming request
URL for the repository event is the same URL we want to use for the
webhook on pull_request events. This simplifies configuration by assuming
that the incoming URL should the URL used for creating the webhook in
hookit().

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd committed:
enable travis testing now that there are tests

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd committed:
provide default empty request signature if none present

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd committed:
correct the expected form of the request header

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd committed:
provide certs to running environment so https succeeds

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd committed:
add a ping hook handler to make adding the hook a bit happier

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd committed:
update readme with new token privilege necessary for webhook creation

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd commented on habitat/hooks/run:
```diff
@@ -0,0 +1,14 @@
+cd /hab/svc/dcob
+
+export RACK_ENV=production
+export SSL_CERT_FILE="$(hab pkg path core/cacerts)/ssl/certs/cacert.pem"
```
Makes me weepy to do this now. I suppose if there is a way to write out our own run hook in some phase of the plan, we could generate this at build time.

robbkidd replied:
Gonna roll this back to a `pkg_svc_run` command in the plan and make setting these a requirement prior to running `hab start chef/dcob`.

robbkidd commented:
Got the bot working on my test org. I'm really to review whenever.

robbkidd commented on src/lib/dcob/octoclient.rb:
```diff
@@ -0,0 +1,21 @@
+                      secret: SECRET_TOKEN }
+      hook_options = { events: ["pull_request"],
+                       active: true }
+      Octokit.create_hook(repository, github_service_name, hook_config, hook_options)
```
Saw a timing problem with the create_hook being possibly being attempted before the repository was fully created at GitHub causing this to fail. Add a retry loop. 😢

robbkidd committed:
add a simple retry loop for adding webhook to repos

Includes a growing sleep interval for each retry.

Refactored the Octokit client into a utility instance method
on Octoclient for clarity in testing.

Signed-off-by: Robb Kidd <robb@thekidds.org>

robbkidd commented:
Created a few repositories in the test org. All successfully received the repo webhook. I don't know whether retries were involved, though, because I haven't sorted out bot logging yet. I supposed that might be next.

thommay reviewed (approved):
Once that hook is gone I'm 👍

robbkidd committed:
move the run hook back into the plan

I created a separate run hook script so that I could set the
production runtime variables for where to fit SSL certs and
to switch the Rack app to production mode. After some discussion,
Thom May and I agreed that those should be concerns for how one
decides to start the bot (e.g. in the systemd or upstart job
definition that starts the hab supervisor).

Updated README with requirement to set production env vars
in whatever is responsible for starting the supervisor.

Signed-off-by: Robb Kidd <robb@thekidds.org>
